### DISCUSSION TITLE
How to install on Ubuntu 22.04?

### DISCUSSION THREAD

**Discussion by johnmee** [qtile/qtile#3558] [Q&A] [answered]
How do I get qtile to run on Ubuntu 22.04?

Apologies if this is a recurring question but, in my defense, the documentation is frustrating.

I have followed the [documentation for Ubuntu](https://docs.qtile.org/en/latest/manual/install/ubuntu.html), 
which fires me off to [install by source](https://docs.qtile.org/en/latest/manual/install/index.html#installing-from-source), then proceeds to discuss options so many years out of date it will apply to 0.00001% of users.

I believe I have managed to install the required dependencies as I am able to run 'qtile start' from a terminal window.  However I am doing that from inside gnome, so it raises 'ExistingWMException' and exits.

After much head scratching I found [at the bottom of the configuration instructions](https://docs.qtile.org/en/latest/manual/config/index.html#starting-qtile) some mention of a desktop file in `/usr/share/xsessions/qtile.desktop`.  This does give me a login profile (in the settings wheel at login) for qtile.  Is that GDM?

At this point I am stuck.  It fails to login (does something then throws me back to the login).  I have not been able to work out where any log files might be, for a clue as to what the problem is.  There is a passing mention of xinitrc which I don't really appreciate.

I'm well versed in python and wanted to give it a try but, so far, the install has beaten me.

I've tried a few 'Exec' variations in the desktop file.  The current of which is `Exec=python3 /home/john/.local/bin/qtile start`

**Comment by SweepingBishops** [comment]
`Exec=qtile start`
should be enough(if that doesn't work make sure that `/home/john/.local` is in your path)
> Apologies if this is a recurring question but, in my defense, the documentation is frustrating.

Work is being done to remedy that. I have a PR open to fix at least the install portions(#3512)
The steps I'd recommed to install qtile on Ubuntu would be:
```
pip install xcffib
pip install qtile
```
Then create a `qtile.desktop` file in the `xsessions` directory.
Edit: Also logs are kept in `~/.local/share/qtile/qtile.log`

**Comment by johnmee** [comment, marked as answer]
I'm still failing.  

I have sprayed libs everywhere now by installing via ubuntu packages as well as via pip and `sudo pip` so who knows what lib is running from which install.  Should I start cross-checking version numbers?

The log reveals that it is not finding a function `cairo_xcb_surface_create` inside libcairo.  So, I guess, libcairo is there, but has the api changed?

```
2022-05-16 08:26:31,100 ERROR libqtile start.py:start():L92 Qtile crashed
Traceback (most recent call last):
  File "/home/john/.local/lib/python3.10/site-packages/libqtile/scripts/start.py", line 90, in start
    q.loop()
  File "/home/john/.local/lib/python3.10/site-packages/libqtile/core/manager.py", line 193, in loop
    asyncio.run(self.async_loop())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/home/john/.local/lib/python3.10/site-packages/libqtile/core/manager.py", line 218, in async_loop
    self.load_config(initial=True)
  File "/home/john/.local/lib/python3.10/site-packages/libqtile/core/manager.py", line 135, in load_config
    self._process_screens(reloading=not initial)
  File "/home/john/.local/lib/python3.10/site-packages/libqtile/core/manager.py", line 343, in _process_screens
    scr._configure(self, i, x, y, w, h, grp, reconfigure_gaps=reconfigure_gaps)
  File "/home/john/.local/lib/python3.10/site-packages/libqtile/config.py", line 305, in _configure
    i._configure(qtile, self, reconfigure=reconfigure_gaps)
  File "/home/john/.local/lib/python3.10/site-packages/libqtile/bar.py", line 295, in _configure
    self.drawer.clear(self.background)
  File "/home/john/.local/lib/python3.10/site-packages/libqtile/backend/x11/drawer.py", line 177, in clear
    self._check_xcb()
  File "/home/john/.local/lib/python3.10/site-packages/libqtile/backend/x11/drawer.py", line 122, in _check_xcb
    self._xcb_surface = self._create_xcb_surface()
  File "/home/john/.local/lib/python3.10/site-packages/libqtile/backend/x11/drawer.py", line 88, in _create_xcb_surface
    surface = cairocffi.XCBSurface(
  File "/usr/local/lib/python3.10/dist-packages/cairocffi/xcb.py", line 37, in __init__
    p = cairo.cairo_xcb_surface_create(
AttributeError: cffi library 'libcairo.so.2' has no function, constant or global variable named 'cairo_xcb_surface_create'
john@deskbox:~/.local/share/qtile$ pip install xcffib
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: xcffib in /home/john/.local/lib/python3.10/site-packages (0.11.1)
Requirement already satisfied: cffi>=1.1.0 in /home/john/.local/lib/python3.10/site-packages (from xcffib) (1.15.0)
Requirement already satisfied: six in /usr/lib/python3/dist-packages (from xcffib) (1.16.0)
Requirement already satisfied: pycparser in /home/john/.local/lib/python3.10/site-packages (from cffi>=1.1.0->xcffib) (2.21)
john@deskbox:~/.local/share/qtile$ sudo apt install libpangocairo-1.0-0 
[sudo] password for john: 
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
libpangocairo-1.0-0 is already the newest version (1.50.6+ds-2).
0 to upgrade, 0 to newly install, 0 to remove and 27 not to upgrade.
john@deskbox:~/.local/share/qtile$ sudo apt install python3-cairocffi 
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
python3-cairocffi is already the newest version (1.2.0-2).
0 to upgrade, 0 to newly install, 0 to remove and 27 not to upgrade.
john@deskbox:~/.local/share/qtile$ sudo apt install python3-xcffib
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
python3-xcffib is already the newest version (0.11.1-2).
0 to upgrade, 0 to newly install, 0 to remove and 27 not to upgrade.
john@deskbox:~/.local/share/qtile$ which qtile
/home/john/.local/bin/qtile
```

**Reply by SweepingBishops** [reply]
Try [this](http://docs.qtile.org/en/stable/manual/hacking.html?highlight=logging#cairo-errors)

**Reply by johnmee** [reply]
I finally came back for another go and managed to:

* look for and uninstall the debian packages for python3-xcflib and python3-cairocffi, then
* `sudo python3 -m pip install xcffib`  
   * docs want updating to the new pip form to avoid warnings and perhaps resultant issues
   * we might want to spell out the `sudo` because we need them installed into the global env?
* `sudo python3 -m pip install cairocffi`  - per the troubleshooting, must be done in that order

And now that I'm in, I just need to work out what to press to, at least, open a browser - might want to add that to 'install instructions' just to verify and pull up a "getting-started" page (I had to come back to gnome to write this :-)

If I get into qtile, I might come back and put together a pull request for docs to this effect.

Thanks.

**Reply by SweepingBishops** [reply]
johnmee, good to know that you've finally worked out the issues!
All PRs are welcome :)
> * docs want updating to the new pip form to avoid warnings and perhaps resultant issues

Just update pip, this is a warning from pip, not qtile.
```
pip install -U pip
```

> * we might want to spell out the `sudo` because we need them installed into the global env?

Hmm, will check into that.

> I just need to work out what to press to, at least, open a browser

I don't think we have anything like that right now. With the default keybindings you can open a browser by:
- using a run launcher like `rofi` or `dmenu`
- opening a terminal and running a browser from that
- press `mod+r` and enter the command to open your browser

But I agree it would be a good idea to make a `guess_browser` function just like we have `guess_terminal` now.
